bottomless/src/transaction_cache.rs

```diff
@@ -0,0 +1,159 @@
+            recovery_fpath,
+            cache: Cache::Memory(BTreeMap::new()),
+        }
+    }
+
+    pub async fn insert(&mut self, pgno: u32, page: &[u8]) -> Result<()> {
+        match &mut self.cache {
+            Cache::Memory(map) => {
+                let len = map.len();
+                match map.entry(pgno) {
+                    Entry::Vacant(_) if len > self.swap_after_pages as usize => {
+                        let page_size = self.page_size;
+                        match self.swap().await? {
+                            Cache::Disk { index, file } => {
+                                Self::persist(index, file, pgno, page_size, page).await?;
+                            }
+                            Cache::Memory(map) => {
+                                map.insert(pgno, page.into());
+                            }
+                        }
+                    }
+                    Entry::Vacant(e) => {
+                        e.insert(page.into());
+                    }
+                    Entry::Occupied(mut e) => {
+                        let buf = e.get_mut();
+                        buf.copy_from_slice(page);
+                    }
+                }
+            }
+            Cache::Disk { index, file } => {
+                Self::persist(index, file, pgno, self.page_size, page).await?;
+            }
+        }
+        Ok(())
+    }
+
+    async fn persist(
+        index: &mut BTreeMap<u32, u64>,
+        file: &mut File,
+        pgno: u32,
+        page_size: u32,
+        page: &[u8],
+    ) -> Result<()> {
+        let end = (index.len() as u64) * (page_size as u64);
+        match index.entry(pgno) {
+            Entry::Vacant(e) => {
+                file.seek(SeekFrom::End(0)).await?;
+                file.write_all(page).await?;
+                e.insert(end);
+            }
+            Entry::Occupied(e) => {
+                let offset = *e.get();
+                file.seek(SeekFrom::Start(offset)).await?;
+                file.write_all(page).await?;
+            }
+        }
+        Ok(())
+    }
+
+    /// Swaps current memory cache onto disk.
+    async fn swap(&mut self) -> Result<&mut Cache> {
+        if let Cache::Disk { .. } = self.cache {
+            tracing::trace!("Swap called on cache already using disk space.");
+            return Ok(&mut self.cache); // already swapped
+        }
+        tracing::trace!("Swapping transaction pages to file {}", self.recovery_fpath);
+        let mut index = BTreeMap::new();
+        let result = OpenOptions::new()
+            .create(true)
+            .write(true)
+            .read(true)
+            .truncate(true)
+            .open(&*self.recovery_fpath)
+            .await;
+        match result {
+            Ok(mut file) => {
+                if let Cache::Memory(old) = &self.cache {
+                    let mut end = 0u64;
+                    for (&pgno, page) in old {
+                        file.write_all(page).await?;
```

Comment:
If `write_all` failed, imo we should print a warning and still fall back to using memory. Worst case, it fails, but perhaps we manage to restore successfully

Reply:
I've made a check on file open - since we explicitly require write access, this check should be enough.